### PR TITLE
(#1287) TextEnvelope is a real envelope

### PR DESCRIPTION
**Pull request by victornoel**
This is for #1287.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1458](https://github.com/yegor256/cactoos/issues/1458) doesn't exist, can't set role

**Comment by victornoel** [comment]
@0crat in

File: src/main/java/org/cactoos/text/Synced.java [hunk 1]
```diff
@@ -48,13 +47,14 @@ public Synced(final Text text) {
...
      */
     public Synced(final Text text, final Object lck) {
-        super(new Scalar<String>() {
-            @Override
-            public String value() throws Exception {
-                synchronized (lck) {
-                    return text.asString();
+        super(
+            new TextOf(
+                () -> {
```
Comment (andreoss) [inline src/main/java/org/cactoos/text/Synced.java:52]: @victornoel Can we reused `scalar.Synched` here?

File: src/main/java/org/cactoos/text/TextEnvelope.java [hunk 2]
```diff
@@ -23,71 +23,46 @@
...
     public final String toString() {
-        return new UncheckedText(this).asString();
+        return this.origin.toString();
     }
 
     @Override
-    public final int hashCode() {
-        return new Unchecked<>(this.origin).value().hashCode();
+    public final boolean equals(final Object obj) {
+        return this.origin.equals(obj);
```
Comment (andreoss) [inline src/main/java/org/cactoos/text/TextEnvelope.java:61]: @victornoel Wouldn't this definition of `equals` cause problems? The operation seems to loose symmetric properties. 
For example
`new TextEnvelope(x).equals(x) == true`, but  not necessarily `x.equals(new TextEnvelope(x))`

**Comment by victornoel** [comment]
@0crat status

File: src/main/java/org/cactoos/text/TextEnvelope.java (hunk 2 above)
Comment (victornoel) [reply src/main/java/org/cactoos/text/TextEnvelope.java:61]: @andreoss this is a known limitation of `equals` actually when an interface has many different implementations. This subject has been discussed many times in the EO community, i.e., what does method equals mean for objects (you can search for "equals" in cactoos issues for example).

My take on it is that if you start having multiple implementations of the same interface, then you can't just expect equals to rely on some internal state, and thus, when you implement it, you have to rely on the exposed behaviour of the interface.

In the case of `Text`, it means that all implementations of `Text` should rely on `Text.asString` to implement `equals`, so in that case, the symmetry will be conserved. It is the same for `MapOf`, `IterableOf` and the other collection-related classes for example. Also sometimes it's better not to implement `equals` at all (e.g., `Iterator`).

Comment (victornoel) [reply src/main/java/org/cactoos/text/TextEnvelope.java:61]: @andreoss see for example https://github.com/yegor256/cactoos/issues/1156#issuecomment-511189091 where I made the same kind of comment. Also I started a discussion on all that matter in #1241, don't hesitate to chime in.

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/1458#issuecomment-691693814))

@victornoel This is what I know about this job in C63314D6Z, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#1458](https://github.com/yegor256/cactoos/issues/1458) is [in scope](http://www.zerocracy.com/policy.html#14) for 2hr
  * The role is `REV`
  * The job is assigned to @andreoss[/z](https://www.0crat.com/u/andreoss) for [1hr](http://www.zerocracy.com/policy.html#8)
  * There is no monetary reward attached, it's a [free](http://www.zerocracy.com/policy.html#2) job
  * The job doesn't have any [impediments](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 15 minutes/points
  * These users are banned and won't be assigned:
    * @victornoel[/z](https://www.0crat.com/u/victornoel): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos%231458%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

File: src/main/java/org/cactoos/text/Synced.java (hunk 1 above)
Comment (victornoel) [reply src/main/java/org/cactoos/text/Synced.java:52]: @andreoss good catch

**Comment by victornoel** [comment]
@andreoss see my changes and answers

**Review by andreoss — APPROVED** [review]
@victornoel Looks fine

**Comment by victornoel** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job was finished in 3 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Job audit: performer @andreoss[/z](https://www.0crat.com/u/andreoss)[/z](https://www.0crat.com/u/andreoss) didn't make CR comments

**Pull request merged**